### PR TITLE
Add ExLlamaV2 and ExLlamav2_HF loaders

### PR DESCRIPTION
**Pull request by oobabooga**
https://github.com/turboderp/exllamav2

Status: ~generates text. Consider this a beta test.~ functional.

### Installation

1) Make sure that you have nvcc in your environment. If not, install it with

```
conda install cuda -c nvidia/label/cuda-11.7.1
```

2) Install ninja and exllamav2:

```
pip install -U ninja exllamav2
```

### Usage

I have made this simple test so far:

```
python server.py --model TheBloke_Llama-2-7b-Chat-GPTQ --loader exllamav2
```

### EXL2 models

Some 70b quants in the new EXL2 format can be found here: https://huggingface.co/turboderp.

Here is a test of [turboderp/LLama2-70B-2.5bpw-h6-exl2](https://huggingface.co/turboderp/LLama2-70B-2.5bpw-h6-exl2) on a RTX 3090:

> Output generated in 8.11 seconds (24.65 tokens/s, 200 tokens, context 15, seed 1964973376)
> Output generated in 8.09 seconds (24.71 tokens/s, 200 tokens, context 15, seed 1268446089)
> Output generated in 7.94 seconds (25.19 tokens/s, 200 tokens, context 15, seed 26057855)

The VRAM usage stayed at a constant 23529MiB.

### CODE REVIEW

**Comment by turboderp** [comment]
Not wasting any time, I see. :)

FWIW, the plan is to maintain a PyPI package with (preferably) an option to install the prebuilt C++ extension, so you should be able to just `pip install exllamav2` and then use it in TGW etc. It's already packaged as a library and there's even technically a v0.0.0 on PyPI. Don't know if that matters to how you'd write the plugin.

I would say that adding typical sampling, TFS, mirostat etc. isn't the highest priority for me right now, so plugging ExLlamaV2 into the HF samplers probably makes a lot of sense, like for V1. In any case, let me know if there's anything you'd want from the API in that regard. There are some changes compared to V1, like there isn't a `last_id_only` argument to the model's forward pass, and stuff like that. But all of that can change, of course.

**Comment by oobabooga** [comment]
Using a pip-installable package is probably better than a relative import like I did. Thanks for the heads up, I'll probably change this before merging this PR.

I have just written an ExLlamav2_HF loader with minor changes from ExLlama_HF and everything just worked, including CFG and perplexity evaluation. The latter used to use the `last_id_only` parameter, but it's working fine without it, so the parameter seems to not be needed anymore.

I might have missed some detail, so I'll wait a bit before merging this PR.

**Comment by BadisG** [comment]
@turboderp the difference in outputs between exllama1_hf and exllama2_hf is quite big, and not in a good way, I noticed the outputs in exllama2 to be poorer in quality, did you add more approximations to get the speed increase?
The image below has the same model, same quant, same prompt, same samplers, same seed, and yet the difference in logits is insane
fFn6oYk

**Comment by Chainfire** [comment]
Not really sure if this is the right place, but when connecting [Aider](https://github.com/paul-gauthier/aider/issues/224) to the openai port, only garbage is produced in Aider using ExLlamaV2 (non-hf). V2 works in the webui chat, and the openai port also still works with V1. I don't see any obvious issue when using curl through that same API with V2 to produce a simple hello world either.

**Comment by YakuzaSuske** [comment]
> https://github.com/turboderp/exllamav2
> 
> Status: ~generates text. Consider this a beta test.~ functional.
> 
> ### Installation
> 1. Make sure that you have nvcc in your environment. If not, install it with
> 
> ```
> conda install cuda -c nvidia/label/cuda-11.7.1
> ```
> 
> 2. Install ninja and exllamav2:
> 
> ```
> pip install -U ninja exllamav2
> ```
> 
> ### Usage
> I have made this simple test so far:
> 
> ```
> python server.py --model TheBloke_Llama-2-7b-Chat-GPTQ --loader exllamav2
> ```
> 
> ### EXL2 models
> Some 70b quants in the new EXL2 format can be found here: https://huggingface.co/turboderp.
> 
> Here is a test of [turboderp/LLama2-70B-2.5bpw-h6-exl2](https://huggingface.co/turboderp/LLama2-70B-2.5bpw-h6-exl2) on a RTX 3090:
> 
> > Output generated in 8.11 seconds (24.65 tokens/s, 200 tokens, context 15, seed 1964973376)
> > Output generated in 8.09 seconds (24.71 tokens/s, 200 tokens, context 15, seed 1268446089)
> > Output generated in 7.94 seconds (25.19 tokens/s, 200 tokens, context 15, seed 26057855)
> 
> The VRAM usage stayed at a constant 23529MiB.

Am i doing something wrong? i installed the stuff on the environment but when i go to load a model using the new exllamav2 i get these two errors.
image
image

**Comment by sammcj** [comment]
Nice work!

I'm now able to load in [70b Q2 models](https://huggingface.co/turboderp/LLama2-70B-chat-2.55bpw-h6-exl2/tree/main) on a single RTX3090 - and it's _fast_:

image

**Comment by Asory2010** [comment]
> > https://github.com/turboderp/exllamav2
> > Status: ~generates text. Consider this a beta test.~ functional.
> > ### Installation
> > 
> > 1. Make sure that you have nvcc in your environment. If not, install it with
> > 
> > ```
> > conda install cuda -c nvidia/label/cuda-11.7.1
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > 
> > 2. Install ninja and exllamav2:
> > 
> > ```
> > pip install -U ninja exllamav2
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > ### Usage
> > I have made this simple test so far:
> > ```
> > python server.py --model TheBloke_Llama-2-7b-Chat-GPTQ --loader exllamav2
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > ### EXL2 models
> > Some 70b quants in the new EXL2 format can be found here: https://huggingface.co/turboderp.
> > Here is a test of [turboderp/LLama2-70B-2.5bpw-h6-exl2](https://huggingface.co/turboderp/LLama2-70B-2.5bpw-h6-exl2) on a RTX 3090:
> > > Output generated in 8.11 seconds (24.65 tokens/s, 200 tokens, context 15, seed 1964973376)
> > > Output generated in 8.09 seconds (24.71 tokens/s, 200 tokens, context 15, seed 1268446089)
> > > Output generated in 7.94 seconds (25.19 tokens/s, 200 tokens, context 15, seed 26057855)
> > 
> > 
> > The VRAM usage stayed at a constant 23529MiB.
> 
> Am i doing something wrong? i installed the stuff on the environment but when i go to load a model using the new exllamav2 i get these two errors. image image

same.... have you found a fix?

**Comment by YakuzaSuske** [comment]
> same.... have you found a fix?

Nope not yet...

**Comment by BadisG** [comment]
> @turboderp the difference in outputs between exllama1_hf and exllama2_hf is quite big, and not in a good way, I noticed the outputs in exllama2 to be poorer in quality, did you add more approximations to get the speed increase? The image below has the same model, same quant, same prompt, same samplers, same seed, and yet the difference in logits is insane [fFn6oYk](https://user-images.githubusercontent.com/110173477/267441342-ec0b6f35-7a99-4bbf-9c9e-94a3589ef729.png)
> 
> Here's some results with the samplers disabled [image](https://user-images.githubusercontent.com/110173477/267447127-bb281d22-8534-4736-ac95-39e1d2470129.png)

I did some perplexity test and exllama2 has worse perplexity aswell.
zrz

**Comment by ThisIsNetsu** [comment]
> > > https://github.com/turboderp/exllamav2
> > > Status: ~generates text. Consider this a beta test.~ functional.
> > > ### Installation
> > > 
> > > 1. Make sure that you have nvcc in your environment. If not, install it with
> > > 
> > > ```
> > > conda install cuda -c nvidia/label/cuda-11.7.1
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > 
> > > 2. Install ninja and exllamav2:
> > > 
> > > ```
> > > pip install -U ninja exllamav2
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > ### Usage
> > > I have made this simple test so far:
> > > ```
> > > python server.py --model TheBloke_Llama-2-7b-Chat-GPTQ --loader exllamav2
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > ### EXL2 models
> > > Some 70b quants in the new EXL2 format can be found here: https://huggingface.co/turboderp.
> > > Here is a test of [turboderp/LLama2-70B-2.5bpw-h6-exl2](https://huggingface.co/turboderp/LLama2-70B-2.5bpw-h6-exl2) on a RTX 3090:
> > > > Output generated in 8.11 seconds (24.65 tokens/s, 200 tokens, context 15, seed 1964973376)
> > > > Output generated in 8.09 seconds (24.71 tokens/s, 200 tokens, context 15, seed 1268446089)
> > > > Output generated in 7.94 seconds (25.19 tokens/s, 200 tokens, context 15, seed 26057855)
> > > 
> > > 
> > > The VRAM usage stayed at a constant 23529MiB.
> > 
> > 
> > Am i doing something wrong? i installed the stuff on the environment but when i go to load a model using the new exllamav2 i get these two errors. image image
> 
> same.... have you found a fix?

Same problem here.

**Comment by YakuzaSuske** [comment]
> > > > https://github.com/turboderp/exllamav2
> > > > Status: ~generates text. Consider this a beta test.~ functional.
> > > > ### Installation
> > > > 
> > > > 1. Make sure that you have nvcc in your environment. If not, install it with
> > > > 
> > > > ```
> > > > conda install cuda -c nvidia/label/cuda-11.7.1
> > > > ```
> > > > 
> > > > 
> > > >     
> > > >       
> > > >     
> > > > 
> > > >       
> > > >     
> > > > 
> > > >     
> > > >   
> > > > 
> > > > 2. Install ninja and exllamav2:
> > > > 
> > > > ```
> > > > pip install -U ninja exllamav2
> > > > ```
> > > > 
> > > > 
> > > >     
> > > >       
> > > >     
> > > > 
> > > >       
> > > >     
> > > > 
> > > >     
> > > >   
> > > > ### Usage
> > > > I have made this simple test so far:
> > > > ```
> > > > python server.py --model TheBloke_Llama-2-7b-Chat-GPTQ --loader exllamav2
> > > > ```
> > > > 
> > > > 
> > > >     
> > > >       
> > > >     
> > > > 
> > > >       
> > > >     
> > > > 
> > > >     
> > > >   
> > > > ### EXL2 models
> > > > Some 70b quants in the new EXL2 format can be found here: https://huggingface.co/turboderp.
> > > > Here is a test of [turboderp/LLama2-70B-2.5bpw-h6-exl2](https://huggingface.co/turboderp/LLama2-70B-2.5bpw-h6-exl2) on a RTX 3090:
> > > > > Output generated in 8.11 seconds (24.65 tokens/s, 200 tokens, context 15, seed 1964973376)
> > > > > Output generated in 8.09 seconds (24.71 tokens/s, 200 tokens, context 15, seed 1268446089)
> > > > > Output generated in 7.94 seconds (25.19 tokens/s, 200 tokens, context 15, seed 26057855)
> > > > 
> > > > 
> > > > The VRAM usage stayed at a constant 23529MiB.
> > > 
> > > 
> > > Am i doing something wrong? i installed the stuff on the environment but when i go to load a model using the new exllamav2 i get these two errors. image image
> > 
> > 
> > same.... have you found a fix?
> 
> Same problem here.

I just fixed my issue. You need visual build tools
https://visualstudio.microsoft.com/visual-cpp-build-tools/
image
install this
image
and that should fix it.

**Comment by YakuzaSuske** [comment]
then just try to update the ui again in case. or restart your pc.

**Comment by ThisIsNetsu** [comment]
How on earth do you get 25 tokens/s?

I only get anywhere from 3-5 tokens/s on a 4090 with the same model used in this post.

**Comment by noobmaster29** [comment]
Yeah I'm also getting the DLL load failed error. I've installed g++ (conda install -c conda-forge m2w64-gcc) and nvcc is on 12.2. Running on windows.

**Comment by JackCloudman** [comment]
I tried to load 70B on RTX4090 but got OOM, i tested [turboderp/LLama2-70B-2.5bpw-h6-exl2](https://huggingface.co/turboderp/LLama2-70B-2.5bpw-h6-exl2)

**Comment by jfan3** [comment]
> Yeah I'm also getting the DLL load failed error. I've installed g++ (conda install -c conda-forge m2w64-gcc) and nvcc is on 12.2. Running on windows.

Have you resolved this? I ran into the same issue... nvcc 12.3, gcc 5.3
